### PR TITLE
#12897 - add mastodon provider

### PR DESCRIPTION
**Pull request by divbhasin**
This PR targets #12897. I included the Mastodon provider in oauth2.go and I am looking for feedback on if I have added (and vendored) the module correctly. The diff seems to be very large and I am wondering if that's normal when adding a new module.

### CODE REVIEW

**Comment by lafriks** [comment]
You need to add changes also in:
models/oauth2.go
templates/admin/auth/new.tmpl
and also add mastodon icon

**Comment by techknowlogick** [comment]
Awesome!! Thanks for the PR :)

A few things:
* Instead of using `Mastodon.New`, you'll need to use `Mastodon.NewCustomisedURL` as otherwise it'll default to mastodon.social and users may want to login with other instances
* You'll need to update some templates as well, to 1. allow user to specify mastodon instance they'd like to use 2. to add it to dropdown
* Here is an example of a PR that changes similar things: https://github.com/go-gitea/gitea/pull/10562

**Comment by divbhasin** [comment]
@techknowlogick I have applied all the suggestions except handling of a custom instance URL. I see that for providers with the custom URL mapping enabled, we have specific endpoints you can customize like the authURL and tokenURL. For mastodon, should we not do any of that and just have the ability to specify a custom instanceURL i.e. something like this:
Screen Shot 2020-10-24 at 3 24 36 PM

**Comment by techknowlogick** [comment]
> should we not do any of that and just have the ability to specify a custom instanceURL i.e. something like this:
Yup :)

> I will also have to another field to the CustomURLMapping struct for InstanceURL, correct?
You could use an existing field (I recommend using AuthURL), but pass it to `Mastodon.NewCustomisedURL` as instanceURL.

File: modules/auth/oauth2/oauth2.go [hunk 1]
```diff
@@ -215,7 +215,13 @@ func createProvider(providerName, providerType, clientID, clientSecret, openIDCo
 		// See https://tech.yandex.com/passport/doc/dg/reference/response-docpage/
 		provider = yandex.New(clientID, clientSecret, callbackURL, "login:email", "login:info", "login:avatar")
 	case "mastodon":
-		provider = mastodon.New(clientID, clientSecret, callbackURL)
+		instanceURL := mastodon.InstanceURL
+		if customURLMapping != nil {
```
Comment (lafriks) [inline modules/auth/oauth2/oauth2.go]: Just use && instead of two if's

**Comment by divbhasin** [comment]
@techknowlogick unable to add you as reviewer but please take a look. Also, is drone down? The automated check doesn't seem to be running.

File: web_src/js/index.js [hunk 2]
```diff
@@ -1831,6 +1832,10 @@ function initAdmin() {
           $('.oauth2_token_url, .oauth2_auth_url, .oauth2_profile_url').show();
           $('#oauth2_email_url').val('');
           break;
+        case 'mastodon':
+          $('.oauth2_auth_url input').attr('required', 'required');
+          $('.oauth2_auth_url').show()
```
Comment (techknowlogick) [inline web_src/js/index.js]:
```suggestion
          $('.oauth2_auth_url').show();
```
frontend lint fails otherwise.

**Review by lafriks — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**